miasm2/ir/ssa.py

```diff
@@ -339,3 +339,267 @@ def transform(self, path):
+        return m2_expr.ExprAff(v, phi)
+
+    @staticmethod
+    def _fill_phi(*args):
+        """
+        Fills a phi function with variables.
+
+        phi(x.1, x.5, x.6)
+
+        :param args: list of ExprId
+        :return: ExprOp
+        """
+        return m2_expr.ExprOp("phi", *args)
+
+    def _transform_phi_rhs(self, src):
+        """
+        Transforms an expression of a phi function on the
+        right hand side into SSA
+        :param src: expression of a phi function on RHS
+        :return: expression in SSA form
+        """
+        # transform variable on RHS in non-SSA form
+        e = self.reverse_variable(src)
+
+        # transform into SSA form
+        src_ssa = self._transform_expression_rhs(e)
+
+        return src_ssa
+
+    def _rename(self, head):
+        """
+        Transforms each variable expression in the CFG into SSA
+        by traversing the dominator tree in depth-first search.
+
+        1. Transform variables of phi functions on LHS into SSA
+        2. Transform all non-phi expressions into SSA
+        3. Update the successor's phi functions' RHS with current SSA variables
+        4. Save current SSA variable stack for successors in the dominator tree
+
+        Source: Cytron, Ron, et al.
+        "An efficient method of computing static single assignment form"
+        Proceedings of the 16th ACM SIGPLAN-SIGACT symposium on
+        Principles of programming languages (1989), p. 31
+        """
+        # compute dominator tree
+        dominator_tree = self.graph.compute_dominator_tree(head)
+
+        # init SSA variable stack
+        stack = [self._stack_rhs.copy()]
+
+        # walk in DFS over the dominator tree
+        for block in dominator_tree.walk_depth_first_forward(head):
+            # restore SSA variable stack of the predecessor in the dominator tree
+            self._stack_rhs = stack.pop().copy()
+
+            '''Transform variables of phi functions on LHS into SSA'''
+            self._rename_phi_lhs(block)
+
+            '''Transform all non-phi expressions into SSA'''
+            self._rename_expressions(block)
+
+            '''Update the successor's phi functions' RHS with current SSA variables'''
+            # walk over block's successors in the CFG
+            for successor in self.graph.successors_iter(block):
+                self._rename_phi_rhs(successor)
+
+            '''Save current SSA variable stack for successors in the dominator tree'''
+            for successor in dominator_tree.successors_iter(block):
+                stack.append(self._stack_rhs.copy())
+
+    def _rename_phi_lhs(self, block):
+        """
+        Transforms phi function's expressions of an IRA block
+        on the left hand side into SSA
+        :param block: IRA block label
+        """
+        if block in self._phinodes:
+            # create temporary list of phi function assignments for inplace renaming
+            tmp = list(self._phinodes[block])
```

Comment:
You can avoid the `if` statement using `tmp = list(self._phinodes.get(block, []))`